sys/net/application_layer/ng_zep/ng_zep.c

```diff
@@ -0,0 +1,1067 @@
+            else {
+                DEBUG("zep: Build ZEPv2 data header in pktbuf\n");
+                zep = ng_pktbuf_add(NULL, NULL, sizeof(ng_zep_v2_data_hdr_t) + size,
+                                    NG_NETTYPE_UNDEF);
+            }
+
+            break;
+
+        default:
+            DEBUG("zep: malconfigured version: %" PRIu8 "\n", dev->version);
+            return NULL;
+    }
+
+    return zep;
+}
+
+static inline size_t _zep_hdr_fill_v1(ng_zep_t *dev, ng_zep_v1_hdr_t *hdr,
+                                      size_t payload_len)
+{
+    hdr->version = 1;
+    hdr->chan = dev->chan - NG_ZEP_MIN_CHANNEL;
+    hdr->dev = byteorder_htons(1);
+    hdr->lqi_mode = dev->lqi_mode;
+    hdr->lqi_val = 0xff;                /* TODO: set */
+    memset(hdr->resv, 0, sizeof(hdr->resv));
+    hdr->length = payload_len;
+
+    return sizeof(ng_zep_v1_hdr_t);
+}
+
+static size_t _zep_hdr_fill_v2_data(ng_zep_t *dev, ng_zep_v2_data_hdr_t *hdr,
+                                    size_t payload_len)
+{
+    uint32_t epoch_sec;
+
+    epoch_sec = (uint32_t)time(NULL);
+
+    hdr->version = 2;
+    hdr->type = NG_ZEP_V2_TYPE_DATA;
+    hdr->chan = dev->chan - NG_ZEP_MIN_CHANNEL;
+    hdr->dev = byteorder_htons(1);
+    hdr->lqi_mode = dev->lqi_mode;
+    hdr->lqi_val = 0xff;                /* TODO: set */
+    hdr->time.b32[0] = byteorder_htonl(epoch_sec);
+    hdr->time.u32[1] = 0;
+    hdr->seq = byteorder_htonl(dev->seq);
+    memset(hdr->resv, 0, sizeof(hdr->resv));
+    hdr->length = payload_len;
+
+    return sizeof(ng_zep_v2_data_hdr_t);
+}
+
+static size_t _zep_hdr_fill(ng_zep_t *dev, ng_zep_hdr_t *hdr,
+                            size_t payload_len)
+{
+    hdr->preamble[0] = 'E';
+    hdr->preamble[1] = 'X';
+
+    switch (dev->version) {
+        case 1:
+            return _zep_hdr_fill_v1(dev, (ng_zep_v1_hdr_t *)hdr, payload_len);
+
+        case 2:
+            return _zep_hdr_fill_v2_data(dev, (ng_zep_v2_data_hdr_t *)hdr,
+                                         payload_len);
+            break;
+
+        default:
+            return 0;
+    }
+}
+
+static ng_pktsnip_t *_create_received(ng_zep_t *dev, ng_pktsnip_t *pkt,
+                                      uint8_t lqi, uint8_t frame_len,
+                                      uint8_t version)
+{
+    ng_pktsnip_t *payload, *mhr, *netif;
+    size_t mhr_len;
+
+    (void)version;
+
+    if ((frame_len != pkt->size) || (_calc_fcs(0, pkt->data, pkt->size) != 0)) {
+        return NULL;
+    }
+
+    payload = ng_pktbuf_add(pkt, pkt->data, pkt->size - 2, dev->proto);
+
+    if (payload == NULL) {
+        return NULL;
+    }
+
+    pkt = ng_pktbuf_remove_snip(pkt, pkt);  /* remove FCS */
+
+    mhr_len = _get_frame_hdr_len(pkt->data);
+
+    if (mhr_len == 0) {
+        return NULL;
+    }
+
+    mhr = ng_pktbuf_add(pkt, pkt->data, mhr_len, NG_NETTYPE_UNDEF);
+
+    /* TODO: send ACK */
+
+    netif = _make_netif_hdr(mhr->data);
+
+    pkt = ng_pktbuf_remove_snip(pkt, mhr);
```

Comment:
Since there are two versions of how to read your sentence:

1. *(Why is MHR removed again here)* It isn't, it is removed for the first time here.
2. *(I don't understand, why do you remove mhr)* Upper layers expect the general netif header, not the link-layer's specific header. The netif header is added in l738, the specific link-layer header (the MHR) is removed here.